app/src/logging/formatters.py

```diff
@@ -6,22 +6,69 @@
 
 See https://docs.python.org/3/library/logging.html#formatter-objects
 """
-
 import json
 import logging
-from datetime import datetime
+from datetime import date, datetime
+from decimal import Decimal
+from enum import Enum
+from typing import Any
+from uuid import UUID
 
 import src.logging.decodelog as decodelog
 
 
+def json_encoder(obj: Any) -> Any:
+    """
+    Handle conversion of various types when logs
+    are serialized into JSON. If not specified
+    will attempt to convert using str() on the object
+    """
+
+    # This is the equivalent of many isinstance(obj, <type>) checks
+    match obj:
```

Comment:
since this is such a low level function that can be called dozens of times per log record, and potentially hundreds of thousands of times for a script that processes a lot of data, i'm a little more concerned about performance.

my understanding of `match` is that it essentially is a much more readable way of doing if/then statements, but it's still scales linearly in the number of statements, whereas a lookup table would be constant time (https://betterprogramming.pub/pythons-match-case-is-too-slow-if-you-don-t-understand-it-8e8d0cf927d)

since we're not doing any fancy pattern matching, can we instead index on the type of the object e.g.

```python

identity = lambda x: x
default_encoder = str
ENCODERS_BY_TYPE = defaultdict(lambda: default_encoder, {
    "str": identity,
    "int": identity,
    "float": identity,
    "bool": identity,
    "int": identity,
    "None": identity,
    "datetime": lambda d: d.isoformat(),
    "date": lambda d: d.isoformat(),
    "Decimal": str,
    "UUID": str,
    "Exception": str,
    "Enum": lambda e: e.value,
    "set": lambda s: list(s),
})


def json_encoder(obj):
    encode = ENCODERS_BY_TYPE[type(obj)]
    return encode(obj)
```

i think this will have more predictable constant time performance

Reply:
Good point on the lookup > match performance, I'll just steal exactly what pydantic_encoder did for that

Reply:
oh cool, looks like they did something [very similar](https://github.com/pydantic/pydantic/blob/main/pydantic/json.py#L45-L70) and LoL they even came up with the same variable name haha that's a funny coincidence